File: cobble-binding/cobble-cpp/cpp/write_batch.cpp
#include <cobble/write_batch.hpp>

#include <limits>
#include <utility>

#include "detail/convert.hpp"
#include "detail/impl.hpp"
#include "detail/options.hpp"

namespace cobble {

WriteBatch::WriteBatch() : impl_(std::make_unique<Impl>()) {}
WriteBatch::WriteBatch(WriteBatch&&) noexcept = default;
WriteBatch& WriteBatch::operator=(WriteBatch&&) noexcept = default;
WriteBatch::~WriteBatch() = default;

void WriteBatch::Put(BucketId bucket, BytesView key, ColumnIndex column,
                     BytesView value, const WriteOptions& options) {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState, "WriteBatch has been moved from");
  }
  const auto native_options = detail::ToNative(options);
  ffi::native_write_batch_put(*impl_->native, bucket, detail::RustBytes(key),
                              column, detail::RustBytes(value), native_options);
}

void WriteBatch::Delete(BucketId bucket, BytesView key, ColumnIndex column,
                        const WriteOptions& options) {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState, "WriteBatch has been moved from");
  }
  const auto native_options = detail::ToNative(options);
  ffi::native_write_batch_delete(*impl_->native, bucket, detail::RustBytes(key),
                                 column, native_options);
}

void WriteBatch::Merge(BucketId bucket, BytesView key, ColumnIndex column,
                       BytesView value, const WriteOptions& options) {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState, "WriteBatch has been moved from");
  }
  const auto native_options = detail::ToNative(options);
  ffi::native_write_batch_merge(*impl_->native, bucket, detail::RustBytes(key),
                                column, detail::RustBytes(value), native_options);
}

std::size_t WriteBatch::size() const noexcept {
  if (!impl_) {
    return 0;
  }
  const auto count = ffi::native_write_batch_len(*impl_->native);
  return count > std::numeric_limits<std::size_t>::max()
             ? std::numeric_limits<std::size_t>::max()
             : static_cast<std::size_t>(count);
}

}  // namespace cobble

File: cobble-binding/cobble-cpp/src/write_batch.rs
use cobble_binding::WriteBatch;

use crate::{ffi, options::to_write_options};

pub(crate) struct NativeWriteBatch {
    pub(crate) batch: WriteBatch,
    count: u64,
}

pub(crate) fn native_write_batch_new() -> Box<NativeWriteBatch> {
    Box::new(NativeWriteBatch {
        batch: WriteBatch::new(),
        count: 0,
    })
}
pub(crate) fn native_write_batch_len(batch: &NativeWriteBatch) -> u64 {
    batch.count
}
pub(crate) fn native_write_batch_put(
    batch: &mut NativeWriteBatch,
    bucket: u16,
    key: &[u8],
    column: u16,
    value: &[u8],
    options: &ffi::NativeWriteOptions,
) {
    batch
        .batch
        .put_with_options(bucket, key, column, value, &to_write_options(options));
    batch.count = batch.count.saturating_add(1);
}
pub(crate) fn native_write_batch_delete(
    batch: &mut NativeWriteBatch,
    bucket: u16,
    key: &[u8],
    column: u16,
    options: &ffi::NativeWriteOptions,
) {
    batch
        .batch
        .delete_with_options(bucket, key, column, &to_write_options(options));
    batch.count = batch.count.saturating_add(1);
}
pub(crate) fn native_write_batch_merge(
    batch: &mut NativeWriteBatch,
    bucket: u16,
    key: &[u8],
    column: u16,
    value: &[u8],
    options: &ffi::NativeWriteOptions,
) {
    batch
        .batch
        .merge_with_options(bucket, key, column, value, &to_write_options(options));
    batch.count = batch.count.saturating_add(1);
}
